Add the `array_histogram` operator

nathanwilk7 opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->

<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description

Adds the `array_histogram` function along with tests and documentation for it.

Fixes https://github.com/trinodb/trino/issues/14725

<!-- Provide details that would help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues

The `array_histogram` function was requested in [issue 14725](https://github.com/trinodb/trino/issues/14725). `martint` suggested the name `array_histogram` to make it consistent with the existing `histogram` function. I've implemented this as a scalar function with the behavior shown below (the input to the function is an array and the output is a map, nulls are ignored).

```sql
SELECT array_histogram(a) FROM (VALUES (ARRAY[1, 2, 1]), (ARRAY[42, 7, 42, null])) t(a);
    _col0
-------------
 {1=2, 2=1}
 {42=2, 7=1}
```

This is my first attempt to contribute to Trino (got the CLA confirmation email and org invite yesterday), feedback is very welcome and I will happily address it, I'm eager to understand the codebase more (this PR was essentially just pattern matching, so I expect there are things I've misused or misunderstood). I've left some comments with questions for reviewers inline as well.

<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# General
* Add the `array_histogram` function. ({issue}`14725 `)
```

nathanwilk7 commented on core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogram.java:
```diff
@@ -0,0 +1,71 @@
+        TypedHistogram histogram = new SingleTypedHistogram(elementType, equalOperator, hashCodeOperator, 10);
+        int positionCount = arrayBlock.getPositionCount();
+        for (int position = 0; position < positionCount; position++) {
+            if (!arrayBlock.isNull(position)) {
```
Do you agree with my choice to ignore nulls? The alternatives would be to either include nulls as a key in the output map with their respective count (assuming null can actually be a key in a map) or to return null as the result of the function if any element in the array is null.

bitsondatadev replied:
I'd say this is a pretty fair assumptions given that's how most of the [other aggregate operators work](https://trino.io/docs/current/functions/aggregate.html).

> Except for [count()](https://trino.io/docs/current/functions/aggregate.html#count), [count_if()](https://trino.io/docs/current/functions/aggregate.html#count_if), [max_by()](https://trino.io/docs/current/functions/aggregate.html#max_by), [min_by()](https://trino.io/docs/current/functions/aggregate.html#min_by) and [approx_distinct()](https://trino.io/docs/current/functions/aggregate.html#approx_distinct), all of these aggregate functions ignore null values and return null for no input rows or when all values are null. For example, [sum()](https://trino.io/docs/current/functions/aggregate.html#sum) returns null rather than zero and [avg()](https://trino.io/docs/current/functions/aggregate.html#avg) does not include null values in the count. The coalesce function can be used to convert null into zero.

nathanwilk7 replied:
Just found the code which does not allow nulls as a key in maps in Trino: https://github.com/trinodb/trino/blob/19f61fc56984bd98c617d8e72214a61237f8b256/core/trino-main/src/main/java/io/trino/operator/scalar/MapConstructor.java#L135

nathanwilk7 commented on core/trino-main/src/test/java/io/trino/operator/scalar/TestArrayHistogram.java:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    @Test
+    public void testTypes()
```
This test is probably overkill, I've tested many of the different types available in Trino, but we could probably strip this down to only a few types. Let me know if you want it kept as is or if you want types cut from the test. The test takes just over 1 second to run on my box.

bitsondatadev replied:
IMO the more the merrier :). This is a good representation of what's possible and a good test to glean examples from.

nathanwilk7 commented on docs/src/main/sphinx/functions/array.rst:
```diff
@@ -60,6 +60,10 @@ Array functions
 
     Returns an array of elements in ``x`` but not in ``y``, without duplicates.
 
+.. function:: array_histogram(x) -> map(K, bigint)
```
This return type of `map(K, bigint)` is not rendering for me in safari or chrome. The same issue happens for the existing `histogram` function from the [Aggregate functions page](https://trino.io/docs/current/functions/aggregate.html#map-aggregate-functions) (shown below in a screenshot). I will look into this rendering issue separate from this PR.
<img width="577" alt="image" src="https://user-images.githubusercontent.com/10050520/217455268-8cf09a6d-5656-494a-bb76-9685e21b00a6.png">

bitsondatadev replied:
This is likely something me and @mosabua need to look into. Please create another issue around this and tag us in it please!

nathanwilk7 commented:
@bitsondatadev @findinpath Would either of you be interested in reviewing this? Or if not, would you be able to suggest someone who could do so?

nathanwilk7 reviewed (commented):
Leaving questions for reviewers

bitsondatadev commented on core/trino-main/src/test/java/io/trino/operator/scalar/TestArrayHistogram.java:
```diff
@@ -0,0 +1,152 @@
+                .binding("a", "ARRAY[42, null]"))
+                .isEqualTo(Map.of(42, 1L));
+    }
+
```
Perhaps we should test negatives, zeros, and other odd values.

nathanwilk7 commented:
@bitsondatadev Per your request, I added some more edge case tests like negative numbers, zero, IP addresses, and timestamps with time zones.

mosabua commented:
Given that this is a new function and hence part of the language, this will need a review from @martint

findinpath commented:
Please add "Fixes #14725" to the description of the ticket in order to eventually close the initial issue once this PR lands.

findinpath commented on core/trino-main/src/test/java/io/trino/operator/scalar/TestArrayHistogram.java:
```diff
@@ -0,0 +1,172 @@
+    public void testNullsIgnored()
+    {
+        assertThat(assertions.expression("array_histogram(a)")
+                .binding("a", "ARRAY[null]"))
```
nit: Use `NULL` in SQL code instead of `null`.

findinpath commented on core/trino-main/src/test/java/io/trino/operator/scalar/TestArrayHistogram.java:
```diff
@@ -0,0 +1,172 @@
+    @Test
+    public void testNullsIgnored()
+    {
+        assertThat(assertions.expression("array_histogram(a)")
```
add an extra assertions with an array containing multiple `NULL` elements.

findinpath commented on core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogram.java:
```diff
@@ -0,0 +1,71 @@
+
+@Description("Return a map containing the counts of the elements in the array")
+@ScalarFunction(value = "array_histogram")
+public final class ArrayHistogram
```
Change class name to `ArrayHistogramFunction`.

findinpath commented on core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogram.java:
```diff
@@ -0,0 +1,71 @@
+                    argumentTypes = "T",
+                    convention = @Convention(arguments = BLOCK_POSITION, result = FAIL_ON_NULL))
+            BlockTypeOperators.BlockPositionHashCode hashCodeOperator,
+            @TypeParameter("map(T, BIGINT)") MapType mapType,
```
`map(T, BIGINT)` -> `map(T, bigint)`

nathanwilk7 commented:
@findinpath Feedback addressed, ready for re-review at your convenience.

Let me know if I should be doing commits differently, it's not clear to me if I should be autosquashing and force pushing or just leaving the fixup commits.

findinpath commented:
Please squash the commits

findinpath commented on core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogramFunction.java:
```diff
@@ -0,0 +1,71 @@
+            @TypeParameter("map(T, bigint)") MapType mapType,
+            @SqlType("array(T)") Block arrayBlock)
+    {
+        TypedHistogram histogram = new SingleTypedHistogram(elementType, equalOperator, hashCodeOperator, 10);
```
Let's use a constant for `10`.

findinpath commented on core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogramFunction.java:
```diff
@@ -0,0 +1,71 @@
+                histogram.add(position, arrayBlock, 1L);
+            }
+        }
+        BlockBuilder blockBuilder = mapType.createBlockBuilder(null, 1);
```
We could add `getPositionCount() {return values.getPositionCount();}` to `SingleTypedHistogram` to avoid expanding on the fly the map block builder internally during the `serialize` operation.

findinpath commented on core/trino-main/src/test/java/io/trino/operator/scalar/TestArrayHistogramFunction.java:
```diff
@@ -0,0 +1,176 @@
+    }
+
+    @Test
+    public void testTypes()
```
You could use `org.testng.annotations.DataProvider` to avoid repeating the same query over and over.

martint replied:
No, don’t use that. It makes it a lot harder to set breakpoints and debug when there are issues.

martint commented on core/trino-main/src/test/java/io/trino/operator/scalar/TestArrayHistogramFunction.java:
```diff
@@ -0,0 +1,176 @@
+                             .binding("a", "ARRAY[42]"))
+                .isEqualTo(Map.of(42, 1L));
+
+        assertThat(assertions.expression("array_histogram(a)")
```
Use assertions.function(), which provides a shortcut to avoid having to use binding()

nathanwilk7 commented on core/trino-main/src/test/java/io/trino/operator/aggregation/TestTypedHistogram.java:
```diff
@@ -31,25 +33,24 @@
 
-        BlockTypeOperators blockTypeOperators = new BlockTypeOperators(new TypeOperators());
         TypedHistogram typedHistogram = new SingleTypedHistogram(
                 BIGINT,
-                blockTypeOperators.getEqualOperator(BIGINT),
-                blockTypeOperators.getHashCodeOperator(BIGINT),
+                BLOCK_TYPE_OPERATORS.getEqualOperator(BIGINT),
+                BLOCK_TYPE_OPERATORS.getHashCodeOperator(BIGINT),
                 1000);
         IntStream.range(1, 2000)
                 .flatMap(i -> IntStream.iterate(i, IntUnaryOperator.identity()).limit(i))
                 .forEach(j -> BIGINT.writeLong(inputBlockBuilder, j));
 
         Block inputBlock = inputBlockBuilder.build();
-        for (int i = 0; i < inputBlock.getPositionCount(); i++) {
-            typedHistogram.add(i, inputBlock, 1);
-        }
+        addInputBlockToTypedHistogram(typedHistogram, inputBlock);
```
Just DRYing up an existing test with the new ones I added below

nathanwilk7 reviewed (commented):
Notes for reviewers

nathanwilk7 commented:
@findinpath @martint Addressed feedback, ready for re-review at your convenience.
- `EXPECTED_HISTOGRAM_SIZE` constant for `10`.
- `getPositionCount` method to `SingleTypedHistogram` class.
- Basic tests for new `getPositionCount` method.
- Use `histogram.getPositionCount` in call to `mapType.createBlockBuilder` to avoid on the fly expansion.
- Replace `assertions.expression` with `assertions.function`.

findinpath reviewed (commented):
Build is red.
`TestTypedHistogram.testGetPositionCountEmpty:118 » IllegalArgument expectedSiz`

nathanwilk7 commented:
@findinpath Thanks fixed the build, sorry about that, I didn't test my last addition locally and I should have let the CI tests complete before I pinged you.

findinpath commented:
@martint  CPTAL?

ebyhr commented:
Removed `syntax-needs-review` label as we already agree the function name in #14725

raunaqmorarka commented on core/trino-main/src/test/java/io/trino/operator/aggregation/TestTypedHistogram.java:
```diff
@@ -59,4 +60,85 @@ public void testMassive()
+        assertEquals(typedHistogram.getPositionCount(), 0);
+    }
+
+    private Block inputBlockFromNumbers(List<Integer> numbersToInsert)
+    {
+        return inputBlockFromNumbers(numbersToInsert, numbersToInsert.size());
+    }
+
+    private Block inputBlockFromNumbers(List<Integer> numbersToInsert, int expectedEntries)
+    {
+        BlockBuilder inputBlockBuilder = BIGINT.createBlockBuilder(null, expectedEntries);
+        for (int n : numbersToInsert) {
+            BIGINT.writeLong(inputBlockBuilder, n);
+        }
+        return inputBlockBuilder.build();
+    }
```
You can use `BlockAssertions#createLongsBlock`

nathanwilk7 replied:
Thanks for the suggestion, I removed the `inputBlockFromNumbers` helper functions and replaced them with `BlockAssertions#createLongsBlock`.

nathanwilk7 commented:
Thanks for the review @raunaqmorarka , I'm out of town without my computer for a few weeks but will address your comments after that time. If someone else prefers to make the changes instead of waiting, feel free to do so!

colebow commented:
> Thanks for the review @raunaqmorarka , I'm out of town without my computer for a few weeks but will address your comments after that time. If someone else prefers to make the changes instead of waiting, feel free to do so!

This should work fine! If the stale bot shows back up, we just need to make sure we remove the 'stale' label so it doesn't get auto-closed on us.

nathanwilk7 commented on core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogramFunction.java:
```diff
@@ -0,0 +1,76 @@
+                histogram.add(position, arrayBlock, 1L);
+            }
+        }
+        BlockBuilder blockBuilder = mapType.createBlockBuilder(null, histogram.getPositionCount());
```
Good question. I'll explain my thought process and please let me know anywhere I'm misunderstanding.

My thought was that `expectedEntries` here refers to the number of elements which we expect to insert into the `blockBuilder` being created on this line. Within the `serialize` function on the next line (72), we write each value into the `blockBuilder`. I believe the `getPositionCount` function on the histogram returns the number of keys in the histogram. So given this, my thought was that I can improve the performance of the `blockBuilder` since on line 71 I expect that `getPositionCount` entries will be inserted into the `blockBuilder` in the `serialize` call on line 72.

If I specified `expectedEntries` as `1`, then I expect the `blockBuilder` would need to resize multiple times if there were many keys in the histogram.

Any corrections welcome, thanks!

raunaqmorarka replied:
Looked into it more closely, I wasn't sure whether the expectedEntries propagated into the key and value builder, but it does. So using `histogram.getPositionCount()` is better.

nathanwilk7 commented on core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogramFunction.java:
```diff
@@ -0,0 +1,76 @@
+                    operator = OperatorType.HASH_CODE,
+                    argumentTypes = "T",
+                    convention = @Convention(arguments = BLOCK_POSITION, result = FAIL_ON_NULL))
+            BlockTypeOperators.BlockPositionHashCode hashCodeOperator,
```
Same as comment on `BlockPositionEqual`

nathanwilk7 commented on core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogramFunction.java:
```diff
@@ -0,0 +1,76 @@
+                    operator = OperatorType.EQUAL,
+                    argumentTypes = {"T", "T"},
+                    convention = @Convention(arguments = {BLOCK_POSITION, BLOCK_POSITION}, result = NULLABLE_RETURN))
+            BlockTypeOperators.BlockPositionEqual equalOperator,
```
You're welcome to ignore this if you prefer to leave the code as is rather than take the time to answer my question, but could you give an example of how one could static import `BlockPositionEqual`? I can't find any other examples of doing so by grepping the codebase and `BlockPositionEqual` is a non-static interface.

nathanwilk7 commented:
@raunaqmorarka This PR is ready for your re-review

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/operator/aggregation/histogram/SingleTypedHistogram.java:
```diff
@@ -173,6 +173,11 @@ public boolean isEmpty()
         return values.getPositionCount() == 0;
     }
 
+    public int getPositionCount()
```
this is fine

nathanwilk7 commented:
@raunaqmorarka Made the final import fix. This should be ready to merge, thanks for the help
